### ISSUE TITLE
PayPal Payflow Pro 500 Internal Server Error

### ISSUE THREAD

**Issue by andyworkbase** [magento/magento2#30541]
### Preconditions (*)

1. PHP 7.1.33 / PHP 7.3.16.
2. MySQL 5.7.31 / MariaDB 10.2.21.
3. Clear Magento CE 2.3.3 without sample data is installed (Luma).
4. Production mode is enabled.
5. PayPal Payflow Pro payment method is configured.
6. Debug mode in PayPal Payflow Pro configuration is enabled.
7. Free Shipping method is enabled.
8. Test Category is set up.
9. Simple Products are created and assigned to the Test Category.

### Steps to reproduce (*)

1. Navigate to storefront.
2. Go to the any product page.
3. Click "Add to Cart".
4. Open mini shopping cart and click "Proceed to Checkout".
5. Fill in the all required shipping fields
6. Choose 'Free Shipping' shipping method and click 'Next'.
7. Choose 'PayPal Payflow Pro' payment method.
8. Fill in the all required fields: Credit Card Number, Expiration Date, Card Verification Number.
9. Click 'Place Order' button.

The next actions are performing (in the order of execution):

1. `/V1/carts/mine/set-payment-information/ -> Magento\Checkout\Api\PaymentInformationManagementInterface::savePaymentInformation()`
2. `\Magento\Paypal\Controller\Transparent\RequestSecureToken::execute()`
3. `https://payflowlink.paypal.com/	`
4. `\Magento\Paypal\Controller\Transparent\Response::execute()` - throw <strong>500 Internal Server Error</strong>

![payment_500](https://user-images.githubusercontent.com/11175986/96426361-d0136300-1205-11eb-8b4e-0bd33a222fdd.png)

The last action throws an error. This error doesn't always occur. Once ordered can be executed without issues, next time with this error. A certain pattern is not observed. Magento log files (<i>debug.log, system.log, exception.log</i>) don't show any information related to issue. <i>payment.log</i> included information about payment request/response:
[payment.log](https://github.com/magento/magento2/files/5400608/payment.log)

This error was reproduced on different environments with different configuration:
1. Clear Magento 2.3.3 instance without sample data (Luma, PHP 7.1.33, MySQL 5.7.31)
2. Magento 2.3.3 instance with custom data (custom theme based on Luma, PHP 7.1.33, MySQL 5.7.31):
~ 100 products;
~ 20 categories;
~ 10 customers;
~ 250 orders.

3. Magento 2.3.3 instance with custom data (custom theme based on Luma, AWS, PHP 7.3.16, MariaDB 10.2.21):
~ 22 000 products;
~ 900 categories;
~ 240 000 customers;
~ 690 000 orders.

### Expected result (*)

Expected, when placing an order via Payflow Pro payment method, Magento should not throw <strong>500 Internal Server Error</strong> when trying to performing the action:
`Magento\Paypal\Controller\Transparent\Response::execute()`
The order should be placed and the customer should be redirected to the success page without any issues.
The order should be created with the status 'Processing' and should be displayed in the sales order grid in the backend.

### Actual result (*)

During the placing the order, Magento throws <strong>500 Internal Server Error</strong> when trying to performing the action:
`Magento\Paypal\Controller\Transparent\Response::execute()`.
Loading stuck at the checkout page.
The order is not placed and not created in the backend.
No messages are displayed for the customer.

------

The problem is critical as it prevents place order for all customers

**Comment by andyworkbase** [comment]
@magento I am working on this

**Comment by andyworkbase** [comment]
@magento add to contributors team

**Comment by andyworkbase** [comment]
@magento add to contributors team

**Comment by yatsumaa** [comment]
You might need to check against this:
Since 2.3.3 The Stores > Sales > Sales emails > Send Order Email Copy To field does not allow a space after the comma between two mail addresses anymore.

 and also core_config_data , the value of sales_email/order/copy_to (do not contain comma in between emails)

**Comment by engcom-Bravo** [comment]
Hello @andyworkbase 

Thank you for your report

We have rechecked this issue on the latest 2.4-develop and 2.3.6 composer
The order is placed successfully using _PayPal Payflow Pro_ Payment method and _Free Shipping_ as Shipping method

Can you please confirm that you can reproduce this issue on 2.4-develop or 2.3.6 composer versions?

Thank you in advance

**Comment by andyworkbase** [comment]
@engcom-Bravo 
It appears that the issue has been resolved since version 2.3.5, as on versions higher than or equal to this, the issue cannot be reproduced. The problem was solved by adding an intermediate redirect between PayPal and Magento:
Controller:
`\Magento\Paypal\Controller\Transparent\Redirect`
Layout:
`vendor\magento\module-paypal\view\frontend\layout\transparent_payment_redirect.xml`
Block:
`\Magento\Payment\Block\Transparent\Redirect`
Template:
`vendor\magento\module-payment\view\frontend\templates\transparent\redirect.phtml`

The return URL's in related method were also changed:
`\Magento\Paypal\Model\Payflow\Service\Request\SecureToken`

```
/**
 * Get the Secure Token from Paypal for TR
 *
 * @param Quote $quote
 * @param string[] $urls
 *
 * @return DataObject
 * @throws \Exception
 */
public function requestToken(Quote $quote, array $urls = [])
{
	$this->transparent->setStore($quote->getStoreId());
	$request = $this->transparent->buildBasicRequest();

	$request->setTrxtype(Payflowpro::TRXTYPE_AUTH_ONLY);
	$request->setVerbosity('HIGH');
	$request->setAmt(0);
	$request->setCurrency($quote->getBaseCurrencyCode());
	$request->setCreatesecuretoken('Y');
	$request->setSecuretokenid($this->mathRandom->getUniqueHash());
	$request->setReturnurl($urls['return_url'] ?? $this->url->getUrl('paypal/transparent/redirect'));
	$request->setErrorurl($urls['error_url'] ?? $this->url->getUrl('paypal/transparent/redirect'));
	$request->setCancelurl($urls['cancel_url'] ?? $this->url->getUrl('paypal/transparent/redirect'));
	$request->setDisablereceipt('TRUE');
	$request->setSilenttran('TRUE');

	$this->transparent->fillCustomerContacts($quote, $request);

	$result = $this->transparent->postRequest($request, $this->transparent->getConfig());

	return $result;
}
```

Also was added a plugin to preserve session cookie after submitting POST form from PayPal to related Magento redirect controller which was mention before
`\Magento\Paypal\Plugin\TransparentSessionChecker`

**Comment by engcom-Bravo** [comment]
Thank you @andyworkbase for the quick response.

Glad that the issue is resolved. 

In this case, we have to close the issue because we can not deliver a fix to Magento 2.3.3
Please feel free to comment, or create new tickets according to the [Issue reporting guidelines](https://github.com/magento/magento2/wiki/Issue-reporting-guidelines)
